[Feat] #12 - 온보딩/모임 모집하기 시간표 UI, 기능 구현

Nya128 opened the pull request:
## 🧩 작업 내용
<!-- 작업한 내용을 간단하게 적어주세요! -->
- 시간표 UI
- 온보딩 시간표 기능 (수업 시간표 -> 공강 시간표 변환)
- 온보딩/모집하기 시간표 분리
- 모집하기 시간표 기능 
  - 특정 요일, 공강시간에만 탭 활성화
  - 선택요일 셀 백그라운드 색상 변경
  - 하나의 공강 내에서 시간 선택 시 중간 공백 자동으로 채워짐
  - 근데 중간에 연속적이지 않은 시간의 다른 공강 탭하면, 기존에 선택했던 공강들 삭제됨

## 🪁 PR Point
<!-- 주의할 사항이나 같이 고민해볼 부분, 강조하고 싶은 내용 등을 적어주세요! -->
### 1️⃣ 일단 시간표는 크게 3개로 분류됨
- 온보딩 TimeTable: 수업 시간 입력
- 모임 모집 TimeTable: 선택 요일에만 공강 시간 입력
- 조회용(ViewMode) TimeTable: 탭 기능 ❌. 조회만 가능 
  (지금은 온보딩에서 공강 시간표로 변환한거 보여줄 때 사용됨. 추후에 공백시간표 조회 기능 때도 사용할 예정)
### 2️⃣ 헷갈릴 만한 프로퍼티 설명
<img width="867" alt="image" src="https://github.com/user-attachments/assets/5eea35c0-c7a3-4f37-b5ef-0bf5caba850e" />

- `freeTimeTable` - 외부에서 공강시간표 데이터를 받아온 것
- `selectedDay` - 사용자가 선택한 모임 모집 요일 (선택 요일 활성화됨)
- `selectedTimeRange` - 셀을 선택할 때마다 모임 시작시간, 종료시간이 설정돼야 함. 그 범위를 Tuple로 정의. 
    <하나의 공강 내에서 시간 선택 시 중간 공백 자동으로 채워짐> 기능 구현을 위함.
- `freeTimeIdToCellsMap` - 하나의 공강마다 id 식별자를 부여했음. 각 공강시간에 해당되는 셀들을 Dictionary로 정의. 
    <중간에 연속적이지 않은 시간의 다른 공강 탭하면, 기존에 선택했던 공강들 삭제됨> 기능 구현을 위함.
- `selectedCells` - 시간표셀이 선택됐을 때 해당 배열에 들어감. 하나의 셀이 중복으로 선택될 수 없으므로 Set으로 정의.
- `currentFreeTimeId` - 현재 내가 선택한 셀들의 공강시간 id를 말함. 다른 공강시간 id에 해당되는 셀을 동시에 선택할 수 없기 때문.
   <중간에 연속적이지 않은 시간의 다른 공강 탭하면, 기존에 선택했던 공강들 삭제됨> 기능 구현을 위함.

### 3️⃣ 주석을 일부러 많이 달았어요.
- 워낙 기능이 많아서.. 조금이라도 자세히 주석을 다는 게 이해에 더 도움이 될 것 같았습니당
### 4️⃣ 추후 리팩토링 할 예정
- 시간표가 3개나 되지만... 전체적인 골격은 똑같아서 추상화한 BaseTimeTable을 만들려고 시도했으나 실패. 추후 스프린트에 다시 도전하겠음. 😢

## 📱 스크린샷
<!-- 작업 내용, 스크린 샷(GIF/MP4) 첨부해주세요. -->
|    구현 내용    |   스크린샷   |
| :-------------: | :----------: |
| [온보딩] 수업시간표 -> 공강시간표 변환 | <img src = "https://github.com/user-attachments/assets/088784fc-12c4-4692-8e7b-469d8ba518ea" width ="300">|
| [모임 모집] 공강 시간표 | <img src = "https://github.com/user-attachments/assets/931540b3-4703-460f-9536-c009e173d081" width ="300">|



### 🔗 Issue 
<!-- 생성한 관련 이슈가 있다면 Resolved #이슈번호로 닫아주세요! -->
Resolved #12

Nya128 committed:
[Feat/#12] 모임 등록 시간표 -선택한 요일 및 공강시간에만 탭 활성화, 색상 분기 처리

Nya128 committed:
[Refactor/#12] 평일 요일 String 배열에서 enum으로 리팩토링

Nya128 committed:
[Feat/#12] 선택 요일 셀에 효과 적용

Nya128 committed:
[Feat/#12] 연속적이지 않은 다른 공강시간 탭하면, 기존에 선택했던 공강셀들 선택 해제 기능 추가

- 공강시간 식별자 id 부여해서 구분
- 각 공강시간 id에 속하는 셀들 Dictionary에 정의
- 같은 요일 내에서 다른 공강시간 탭하면 기존 선택셀들 삭제

Nya128 committed:
[Feat/#12] 하나의 공강 내에서 시간 선택 시 중간 공백 자동으로 채워지는 기능 구현

- selectedTime에 튜플로 모임 시작 시간, 종료 시간 저장해서 그 사이 미선택 셀들도 다 선택되게 함

Nya128 committed:
[Refactor/#12] body 내부 뷰들 함수로 추출해 코드 정리 및 전반적인 로직 리팩토링, 문서화 주석 추가

Nya128 committed:
[Feat/#12] 시간표 네이밍 수정 및 폴더링 변경

Nya128 committed:
[Feat/#12] 온보딩 시간표 UI 구현

Nya128 committed:
[Feat/#12] 온보딩 시간표 - 선택된 셀들 수업 시간표로 변경 기능 구현

Nya128 committed:
[Refactor/#12] 상위뷰에서 시간표를 사용하므로 선택 셀들, 수업시간표 데이터 프로퍼티 모두 @State에서 @Binding으로 변경

Nya128 committed:
[Chore/#12] 선택 셀들 수업시간표 모델 데이터로 변환 함수 extension으로 이동

Nya128 committed:
[Feat/#12] 조회용 공강 시간표 따로 만듦

Nya128 committed:
[Feat/#12] 온보딩 시간표 변환 과정 테스트용 예시 코드 작성

Nya128 committed:
[Refactor/#12] 모임 등록 시간표 @Binding 필요한 변수들 정리

Nya128 committed:
[Chore/#12] 프리뷰 예시 코드 수정

Nya128 committed:
Merge branch 'develop' into feat/#12-timetable

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+import SwiftUI
+
+struct AddMeetingTimeTable: View {
+    private let hours = Array(stride(from: 9, through: 17.5, by: 0.5))
```
요거 시간은 시작 시간을 기준으로 하는 건가용? 17.5 = 6시까지? 맞나용?

Nya128 replied:
헤헤 저게 0.5단위로 9시부터 18시까지 배열 만든건데... 종료값인 17.5까지 0.5가 더해지는 건가봐요!!
 지피티한테 자세히 물어봤어요
```
Swift의 stride(from:through:by:) 함수는 일정한 간격으로 연속된 숫자 시퀀스를 생성할 때 사용됩니다.

from: 시작 값 (9)
through: 종료 값 (17.5) → 종료 값을 포함
by: 증가 값 (0.5) → 0.5씩 증가하면서 배열을 만듦
```

HEHEEUN commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+        GridRow {
+            /// 빈칸
+            Rectangle()
+                .fill(Color.white)
```
.grayWhite쓰면 Color.white 안써도 돼용!

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+                /// 선택 요일인지 여부
+                let isSelectedDay = selectedDay == WeekDay.allCases[dayIndex]
+                
+                Text(WeekDay.allCases[dayIndex].rawValue)
```
선택된 요일은 검정배경이도록 하는!

HEHEEUN commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+        
+        return Text(isOnTheHour ? "\(Int(hours[hourIndex]))" : "")
+            .frame(width: 24, height: 28, alignment: .topTrailing)
+            .font(.pretendard(.caption2_r_12))
```
.pretendardFont쓰면 자간,행간도 적용돼요!

Nya128 replied:
허거덩 폰트 대충 적용했더니 완전 잘못 넣어버렸구만! 수정할게유 ❤️

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+        /// 시간을 1로 나눴을 때 나머지 존재 여부 -> 30분 단위 시간 구분 위함
+        let isOnTheHour = hours[hourIndex].truncatingRemainder(dividingBy: 1) == 0
+        
+        return Text(isOnTheHour ? "\(Int(hours[hourIndex]))" : "")
```
정각만 시간표에 나오도록 하는 코드~!

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+    
+    // MARK: - Helper Methods
+    
+    private func bottomPadding(
```
우와 이거 흐린눈 하려고 했는뎅!! 엄마는 진짜 체고다!

Nya128 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ역시 민서는 다 눈에 보였구나... 나도 좀 거슬려서 ㅎ 신경좀 써봤으 😏

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+            )).map {
+                TimeTableCellId(hourIndex: $0, dayIndex: dayIndex)
+            }
+            /// 딕셔너리 key는 공강시간 id, value는 해당 공강시간에 속하는 cells
```
대박 멋쩌,,

HEHEEUN commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/OnboardingTimeTable.swift:
```diff
@@ -0,0 +1,157 @@
+                classTimeTable.append((day: WeekDay.allCases[dayIndex], start: s, end: e))
+            }
+        }
+    }
```
ㄹㅈㄷ 천재인가바요 엄마는

Nya128 replied:
하나둘셋! 지피티야 고마워~ 😎

kms0233 commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+            
+            /// 공강시간 id 같은 경우, 미선택됐던 중간 셀들도 선택하는 기능
+            /// 선택된 셀들의 시작시간 중 최소, 종료시간 중 최대를 튜플로 selectedTimeRange에 저장
+            selectedTimeRange = selectedCells.reduce(into: (
```
선택된 cell들 중 연속된 블록 업데이트

HEHEEUN commented on Gongbaek_iOS/Gongbaek_iOS/Global/Component/Others/TimeTable/AddMeetingTimeTable.swift:
```diff
@@ -0,0 +1,219 @@
+import SwiftUI
+
+struct AddMeetingTimeTable: View {
+    private let hours = Array(stride(from: 9, through: 17.5, by: 0.5))
```
난 stride가 먼지 몰랐어! 엄마 고마워요!

Nya128 replied:
나도 이번에 처음 제대로 써봤어!!

kms0233 reviewed (approved):
![image](https://github.com/user-attachments/assets/22797d95-986a-42e3-89e5-e5084be11804)
엄마가 내 엄마라 너무 행복하다 ㅎㅎ
흐름은 대충 이해했는데 더 자세히 이해하려면.. 직접 써봐야 알 것 같아요
엄마체고 고생했어 헤헤헤😝헤 덕분에 수면 시간 20시간 늘었다 ㅋㅋ

HEHEEUN reviewed (approved):
엄마.. 사실 이해 못했어요..
내 두뇌로는 아직 어려운가봐요
흐름은 대충 알겟어요
주석도 예쁘게 달아줘서 고마워요


![](https://i.pinimg.com/736x/df/6e/bf/df6ebf02fcc7af67fd2c9373eb24c136.jpg)
🩷 울 엄마 시간표 지옥 해방 축하 ! 고생 넘넘 많앗져 🩷

Nya128 committed:
[Chore/#12] 코드리뷰 반영

- 색상 리소스 변경 (white -> grayWhite)
- 폰트 modifier 변경: 자간, 행간 설정된 것 (font -> pretendardFont 변경)
